drivers/flash/flash_stm32_fmc_nand.c

```diff
@@ -0,0 +1,767 @@
+		return -EIO;
+	}
+
+	dev_data->state = NAND_STATE_BUSY;
+	nand_address = STM32_FMC_NAND_ARRAY_ADDRESS(address, dev_data); /* Raw NAND address */
+
+	/* Send read page command sequence */
+	sys_write8(NAND_CMD_AREA_A, NAND_DEVICE | CMD_AREA);
+
+	if (dev_data->page_size <= 512) {
+		if (dev_data->total_pages <= 65535) {
+			sys_write8(0x00, NAND_DEVICE | ADDR_AREA);
+			sys_write8(ADDR_1ST_CYCLE(nand_address), NAND_DEVICE | ADDR_AREA);
+			sys_write8(ADDR_2ND_CYCLE(nand_address), NAND_DEVICE | ADDR_AREA);
+		} else {
+			sys_write8(0x00, NAND_DEVICE | ADDR_AREA);
+			sys_write8(ADDR_1ST_CYCLE(nand_address), NAND_DEVICE | ADDR_AREA);
+			sys_write8(ADDR_2ND_CYCLE(nand_address), NAND_DEVICE | ADDR_AREA);
+			sys_write8(ADDR_3RD_CYCLE(nand_address), NAND_DEVICE | ADDR_AREA);
+		}
+	} else {
+		if (dev_data->total_pages <= 65535) {
+			sys_write8(0x00, NAND_DEVICE | ADDR_AREA);
+			sys_write8(0x00, NAND_DEVICE | ADDR_AREA);
+			sys_write8(ADDR_1ST_CYCLE(nand_address), NAND_DEVICE | ADDR_AREA);
+			sys_write8(ADDR_2ND_CYCLE(nand_address), NAND_DEVICE | ADDR_AREA);
+		} else {
+			sys_write8(0x00, NAND_DEVICE | ADDR_AREA);
+			sys_write8(0x00, NAND_DEVICE | ADDR_AREA);
+			sys_write8(ADDR_1ST_CYCLE(nand_address), NAND_DEVICE | ADDR_AREA);
+			sys_write8(ADDR_2ND_CYCLE(nand_address), NAND_DEVICE | ADDR_AREA);
+			sys_write8(ADDR_3RD_CYCLE(nand_address), NAND_DEVICE | ADDR_AREA);
+		}
+	}
+
+	sys_write8(NAND_CMD_AREA_TRUE1, NAND_DEVICE | CMD_AREA);
+
+	/* Read status until NAND is ready or reports an error */
+	ret = flash_stm32_fmc_nand_wait();
+	if (ret == -EIO) {
+		LOG_ERR("Uncorrectable ECC error detected");
+		dev_data->state = NAND_STATE_READY;
+		k_sem_give(&dev_data->lock);
+		return ret;
+	} else if (ret == -ETIMEDOUT) {
+		dev_data->state = NAND_STATE_ERROR;
+		k_sem_give(&dev_data->lock);
+		return ret;
+	}
+
+	/* Go back to read mode */
+	sys_write8(NAND_CMD_AREA_A, NAND_DEVICE | CMD_AREA);
+
+#if STM32_FMC_NAND_USE_DMA
+	if (dev_data->dma.enabled) {
+		/* Get page data into buffer */
+		if (HAL_DMA_Start(&dev_data->dma.handle, NAND_DEVICE, (uint32_t)config->page_buffer,
+				  dev_data->page_size) != HAL_OK) {
+			dev_data->state = NAND_STATE_ERROR;
+			k_sem_give(&dev_data->lock);
+			return -EIO;
+		}
+		if (HAL_DMA_PollForTransfer(&dev_data->dma.handle, HAL_DMA_FULL_TRANSFER,
+					    DMA_TIMEOUT_MS) != HAL_OK) {
+			dev_data->state = NAND_STATE_ERROR;
+			k_sem_give(&dev_data->lock);
+			return -EIO;
+		}
+
+		/* Get chunk into output buffer */
+		if (dma_reload(dev_data->dma.dev, dev_data->dma.channel,
+			       (uint32_t)(config->page_buffer + page_offset), (uint32_t)data,
+			       chunk) != 0) {
+			dev_data->state = NAND_STATE_ERROR;
+			k_sem_give(&dev_data->lock);
+			return -EIO;
+		}
+	} else {
+#endif /* STM32_FMC_NAND_USE_DMA */
+
+		/* Get page data into buffer */
+		for (size_t index = 0; index < dev_data->page_size; index++) {
+			config->page_buffer[index] = sys_read8(NAND_DEVICE);
+		}
```

Comment:
Using `config->page_buffer` allocated in cache memory is useful when using DMAs, but here, maybe worth the straight load data read into the likely cached `data` buffer.
If using this scheme, you could avoid `flash_stm32_fmc_nand_page_buffer_XXX[]` allocations.